File: 3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.java
class Solution {
    public boolean[] pathExistenceQueries(int n, int[] nums, int maxDiff, int[][] queries) {
        int group[] = new int[n];
        int gno = 0;
        group[0] = 0;
        for(int i=1; i<n; i++) {
            if(nums[i]-nums[i-1] > maxDiff) gno++;
            group[i] = gno;
        }
        
        boolean[] res = new boolean[queries.length];
        int i=0;
        for(int[] que: queries) {
            int u = que[0], v = que[1];
            if(group[u] == group[v]) res[i] = true;
            i++;
        }
        return res;
    }
}

File: 3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.cpp
class Solution {
public:
    vector<int> parent;
    vector<int> size;

    int find(int x) {
        if (parent[x] != x)
            parent[x] = find(parent[x]);
        return parent[x];
    }

    void unite(int x, int y) {
        int px = find(x);
        int py = find(y);

        if (px == py)
            return;
        if (size[px] < size[py]) {
            parent[px] = py;
            size[py] += size[px];
        } else {
            parent[py] = px;
            size[px] += size[py];
        }
    }
    vector<bool> pathExistenceQueries(int n, vector<int>& nums, int maxDiff,
                                      vector<vector<int>>& queries) {
        parent.resize(n);
        size.resize(n, 1);
        for (int i = 0; i < n; i++)
            parent[i] = i;
        for (int i = 0; i < n - 1; i++) {
            if (nums[i + 1] - nums[i] <= maxDiff)
                unite(i, i + 1);
        }
        vector<bool> res;
        for (auto& q : queries) {
            int px = find(q[0]), py = find(q[1]);
            res.push_back(px == py);
        }
        return res;
    }
};